### ISSUE TITLE
Defringe banding

### ISSUE THREAD

**Issue by MinePro120** [Beep6581/RawTherapee#6929]
**Short description**
Cross-post from [here](https://discuss.pixls.us/t/defringe-banding/41564), the defringe tool creates banding visible in cases of extreme purple fringes. CA correction tools do not help, neither does changing the defringe settings. I usually set the threshold to zero to get the smoothest result. The image below is the difference of the demosaiced images, with and without the defringe tool applied. A periodic banding is clearly present.

![diff-screen-scaled](https://github.com/Beep6581/RawTherapee/assets/37949795/2c912d91-341d-4f68-9a25-b28740e2dad4)

**Steps to reproduce**
1. Open the raw image I provided
2. Load the included profile
3. Banding caused by the defringe is most noticeable on the left side of the left bird.

**Expected behavior**
The defringe tool should create smooth results (at least with zero threshold), without creating any visible artifacts or banding.

**Additional information**
- Version of RawTherapee: 5.9
- Version of your operating system: Arch

Other useful information:
[Raw, pp3, diff](https://discuss.pixls.us/uploads/short-url/lMCsMSEWGxArXyyyzfjKDetaZXR.7z)

**Comment by TechXavAL** [comment]
Even though there are still fringes here and there, you may wish to test this pp3:

[photo.NEF.pp3.zip](https://github.com/Beep6581/RawTherapee/files/13998806/photo.NEF.pp3.zip)
RT v.5.9 release *(not tested with latest builds)*.

And yet, with such heavy fringes, sadly you can't expect any miracles.

From my point of view, the defringe tool works as expected, nevertheless it's quite picky in demanding situations.

**Comment by MinePro120** [comment]
Defringe works very well IMO, apart from those bands. Are those really the best result? To me it looks like a bug, or something that can definitely be improved. Using your profile, and by using a defringe radius of 4, the bands are pretty noticeable once again.
![2024-01-20_20-32](https://github.com/Beep6581/RawTherapee/assets/37949795/b0b9074d-d980-4c9f-b5e7-8b635893b02d)

**Comment by TechXavAL** [comment]
Again, you shouldn't expect miracles from this tool: you must do your part helping the tool to achieve its best.

Defringing is basically a desaturation process: you tell the tool how strong, how wide and on which colors it has to desaturate the fringes. You set wrong values, you get artifacts.

You may wish to take a look at [this RawPedia page](http://rawpedia.rawtherapee.com/Defringe/es): even if you don't understand the text, you can clearly see what I'm talking about in the second (comparison) image: left original image, center reasonable settings, right wrong settings (way too high settings).

If I'm not wrong, in your first example the radius was too low, hence the gray fringe below the blue one. The second example is the other way around, that is, the radius is way too high: the gray band is nothing more than the desaturated leg. It's not an artifact, but just a black and white rendition of the leg.

**Comment by MinePro120** [comment]
I have read this page before. I'm well aware defringe causes excessive desaturation if not applied properly, but this is not my issue. In all my examples, the radius is quite high (above 3). But even when set to 3, the bands start becoming visible. There are always there, just less visible. Sure some tools will create artifacts if the settings are weird, but such rough "digital" artifacts seem like a tool's  issue to me. In the wiki, the colour "bleed" is caused by the large radius and the rough edges are caused by the threshold, if I'm not mistaken. I'm not talking about any of those. And no, the desaturated rendition of the leg has no bands. At this point it would be best to get an opinion from someone who knows the code of the tool.

**Comment by Lawrence37** [comment]
There is a bug in the defringe code. It writes the result into the same buffer it's reading from. Since it samples the color from the partially defringed image, it desaturates the image more than it should. The banding is a secondary effect resulting from the parallelization. The image is processed in 16 pixel tall strips, with the top of each strip having less desaturation because the pixels above it haven't been defringed yet.

The following patch fixes the bug for one of the defringing functions by duplicating the buffer to separate the reading and writing.
```diff
diff --git a/rtengine/PF_correct_RT.cc b/rtengine/PF_correct_RT.cc
index 15b7b21ee..772fb3efd 100644
--- a/rtengine/PF_correct_RT.cc
+++ b/rtengine/PF_correct_RT.cc
@@ -134,6 +134,8 @@ void ImProcFunctions::PF_correct_RT(LabImage * lab, double radius, int thresh)
         const float threshfactor = 1.f / (SQR(thresh / 33.f) * chromavef * 5.0f + chromavef);
         const int halfwin = std::ceil(2 * radius) + 1;
 
+        LabImage originalLab(*lab, multiThread);
+
 // Issue 1674:
 // often, colour fringe is not evenly distributed, e.g. a lot in contrasty regions and none in the sky.
 // so it's better to schedule dynamic and let every thread only process 16 rows, to avoid running big threads out of work
@@ -156,8 +158,8 @@ void ImProcFunctions::PF_correct_RT(LabImage * lab, double radius, int thresh)
                         for (int j1 = 0; j1 < j + halfwin; j1++) {
                             // neighbourhood average of pixels weighted by chrominance
                             const float wt = fringe[i1 * width + j1];
-                            atot += wt * lab->a[i1][j1];
-                            btot += wt * lab->b[i1][j1];
+                            atot += wt * originalLab.a[i1][j1];
+                            btot += wt * originalLab.b[i1][j1];
                             norm += wt;
                         }
 
@@ -176,8 +178,8 @@ void ImProcFunctions::PF_correct_RT(LabImage * lab, double radius, int thresh)
                         for (int j1 = j - halfwin + 1; j1 < j + halfwin; j1++) {
                             // neighbourhood average of pixels weighted by chrominance
                             const float wt = fringe[i1 * width + j1];
-                            atot += wt * lab->a[i1][j1];
-                            btot += wt * lab->b[i1][j1];
+                            atot += wt * originalLab.a[i1][j1];
+                            btot += wt * originalLab.b[i1][j1];
                             norm += wt;
                         }
 
@@ -196,8 +198,8 @@ void ImProcFunctions::PF_correct_RT(LabImage * lab, double radius, int thresh)
                         for (int j1 = j - halfwin + 1; j1 < width; j1++) {
                             // neighbourhood average of pixels weighted by chrominance
                             const float wt = fringe[i1 * width + j1];
-                            atot += wt * lab->a[i1][j1];
-                            btot += wt * lab->b[i1][j1];
+                            atot += wt * originalLab.a[i1][j1];
+                            btot += wt * originalLab.b[i1][j1];
                             norm += wt;
                         }
 
```

**Comment by MinePro120** [comment]
Amazing, thanks!
![2024-01-22_10-11](https://github.com/Beep6581/RawTherapee/assets/37949795/3b7ca12a-86f2-4562-8f42-9aea7be07002)

**Comment by MinePro120** [comment]
Can we expect this to make it into 5.10, and if yes, will that be released anytime soon?

**Comment by Lawrence37** [comment]
Maybe, and maybe. The fix is simple enough to squeeze into 5.10. I just need to verify the fix for the other defringing code. The release has been planned for some time now. There is one outstanding issue and some preparations left.
